src/ConnectionPool.jl

```diff
@@ -459,16 +476,48 @@ function keepalive!(tcp)
     return
 end
 
+struct ConnectTimeout <: Exception
+    host
+    port
+end
+
 function getconnection(::Type{Sockets.TCP},
                        host::AbstractString,
                        port::AbstractString;
                        keepalive::Bool=false,
+                       connect_timeout::Int=0,
                        kw...)::Sockets.TCP
+
     p::UInt = isempty(port) ? UInt(80) : parse(UInt, port)
+
     @debug 2 "TCP connect: $host:$p..."
-    tcp = Sockets.connect(Sockets.getaddrinfo(host), p)
-    keepalive && keepalive!(tcp)
-    return tcp
+
+    if connect_timeout == 0
+        tcp = Sockets.connect(Sockets.getaddrinfo(host), p)
+        keepalive && keepalive!(tcp)
+        return tcp
+    end
+
+    tcp = Sockets.TCPSocket()
+    Base.connect!(tcp, Sockets.getaddrinfo(host), p)
+    yield()
+
+    # sum(delays) ~= connect_timeout
+    delays = ExponentialBackOff(n=connect_timeout + 10,
```

Comment:
Why are you busy-polling this? We have a callback for it (wait_connected)

Reply:
I couldn't see how to simultaneously wait for connection success and a timeout.
I can call wait_connected in a background task, and notify a condition when it is done, but there is no timeout parameter wait(::Condition). Or I could wait for the task to be done, but `fetch(::Task)` has no timeout parameter.

Are you thinking along the lines of `wait_connected` in the main task and a background task that calls close after the timeout to wake up the `wait_connected` with an error? I used that pattern over here: https://github.com/JuliaWeb/HTTP.jl/blob/master/src/TimeoutRequest.jl#L20
Don't know why it didn't occur to me in this case :)

Reply:
I think you answered your own question...yes.